zkevm-circuits/src/evm_circuit/execution/precompiles/ec_mul.rs

```diff
@@ -0,0 +1,446 @@
+                    ret_offset: 0x00.into(),
+                    ret_size: 0x40.into(),
+                    address: PrecompileCalls::Bn128Mul.address().to_word(),
+                    ..Default::default()
+                },
+
+                PrecompileCallArgs {
+                    name: "ecMul (valid inputs > 96 bytes)",
+                    // P = (2, 16059845205665218889595687631975406613746683471807856151558479858750240882195)
+                    // s = 7
+                    setup_code: bytecode! {
+                        // p_x
+                        PUSH1(0x02)
+                        PUSH1(0x00)
+                        MSTORE
+
+                        // p_y
+                        PUSH32(word!("0x23818CDE28CF4EA953FE59B1C377FAFD461039C17251FF4377313DA64AD07E13"))
+                        PUSH1(0x20)
+                        MSTORE
+                        
+                        // s
+                        PUSH1(0x07)
+                        PUSH1(0x40)
+                        MSTORE
+
+                        // junk bytes, will be truncated
+                        PUSH32(0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFu128)
+                        PUSH1(0x80)
+                        SHL
+                        PUSH32(0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFu128)
+                        ADD
+                        PUSH1(0x60)
+                        MSTORE
+                    },
+                    call_data_offset: 0x00.into(),
+                    call_data_length: 0x80.into(),
+                    ret_offset: 0x80.into(),
+                    ret_size: 0x40.into(),
+                    address: PrecompileCalls::Bn128Mul.address().to_word(),
+                    ..Default::default()
+                },
+
+                PrecompileCallArgs {
+                    name: "ecMul (invalid input: must mod p to be valid)",
+                    // P = (p + 1, p + 2)
+                    // s = 7
+                    setup_code: bytecode! {
+                        // p_x
+                        PUSH32(word!("0x30644E72E131A029B85045B68181585D97816A916871CA8D3C208C16D87CFD48"))
+                        PUSH1(0x00)
+                        MSTORE
+
+                        // p_y
+                        PUSH32(word!("0x30644E72E131A029B85045B68181585D97816A916871CA8D3C208C16D87CFD49"))
+                        PUSH1(0x20)
+                        MSTORE
+                        
+                        // s = 7
+                        PUSH1(0x07)
+                        PUSH1(0x40)
+                        MSTORE
+                    },
+                    call_data_offset: 0x00.into(),
+                    call_data_length: 0x60.into(),
+                    ret_offset: 0x60.into(),
+                    ret_size: 0x00.into(),
+                    address: PrecompileCalls::Bn128Mul.address().to_word(),
+                    ..Default::default()
+                },
+
+                // TODO: Failing Test
```

Comment:
Test fails with unsatisfied `evm_checks` on the mock prover. 
Test passes if the magnitude of scalar s is reduced. 
For example, ```s = 0x20644E72E131A029B85045B68181585D97816A916871CA8D3C208C16D87CFEE8``` works while ```s = 0x30644E72E131A029B85045B68181585D97816A916871CA8D3C208C16D87CFEE8```  fails.

The gate constraint on offset 62 is the only unsatisfied constraint. All other constraint checks on evm circuit passed. 

To reproduce, uncomment failing cases, then run:
```
RUST_LOG=info cargo test -p zkevm-circuits --features test evm_circuit::execution::precompiles::ec_mul
```
Output:
```
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
left: `Err([ConstraintCaseDebug {
    constraint: Constraint {
        gate: Gate {
            index: 56,
            name: "CALL_OP",
        },
        index: 226,
        name: "output bytes == 0",
    },
    location: InRegion {
        region: Region 12 ('Execution step'),
        offset: 62,
    },
    cell_values: [
        (
            DebugVirtualCell {
                name: "",
                column: DebugColumn {
                    column_type: Advice,
                    index: 60,
                    annotation: "EVM_q_step",
                },
                rotation: 0,
            },
            "1",
        ),
    ],
}])`,
right: `Ok(())`', zkevm-circuits/src/test_util.rs:96:17
```

Reply:
I will come back to this later in the week after I'm done resolving some updates to the ECC Circuit, and `ecPairing` gadget.

Reply:
@darth-cy the error was because:
- Scalar `s` was being created by using the `Fr::from_bytes` method, which expects the input to be in canonical form, i.e. should be less than the modulus. So `s` was `Choice(0)` in this failing case.
- Eventually, `r_expected` did not match `r_specified` and hence `is_valid` was marked as `false` (even though the computation should have been valid). So the `output == 0` constraint failed.
- I've updated that part of the source code to make use of `Fr::from_raw` and supply the little-endian `[u64; 4]` input from reading the 256-bit input as a `U256`.

After making this change, the only error now is failure while doing lookup to the ECC Table.

This is expected, as the ECC Table's assignments are from `EcMulOp` that has fields `(P: (G1_x, G1_y), s, R: (G1_x, G1_y))`. Whereas, the lookup is being done by taking the 256-bit EVM inputs (`s_rlc` is the culprit).

We will need to reduce that 256-bit `s` into `Fr` before passing it in the lookup argument.

Reply:
In order to address this issue, you can have a look at our `ModGadget` in the `evm_circuit::util` module.

We need to reduce the EVM input, that means:
```
evm_input * R == scalar_s (mod r)
```

where:
- `evm_input`: 32-bytes input for `s` from EVM
- `R`: `2 ^ 256 (mod r)`, or `0xe0a77c19a07df2f666ea36f7879462e36fc76959f60cd29ac96341c4ffffffb`
- `scalar_s`: The reduced scalar (this is the entry in the ECC table)
- `r`: Modulus for `Fr`, or `0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001`